File: lib/qa.ts
import "server-only";

import { apiFetch } from "./api/client";

export type QAAuthor = {
  id: string;
  username: string | null;
  display_name: string | null;
  trust_stage: number;
  trust_level_name: string | null;
  /** 0..100 (ADR-003); shown beside the level name. See lib/trust.ts. */
  reputation_score: string | null;
};

export type Answer = {
  id: string;
  body: string;
  is_best_answer: boolean;
  is_first_responder: boolean;
  helpful_votes: number;
  created_at: string;
  responder: QAAuthor | null;
};

export type Question = {
  id: string;
  product_id: string;
  product_name: string | null;
  body: string;
  directed_to: "buyers" | "seller";
  best_answer_id: string | null;
  answer_count: number;
  created_at: string;
  asker: QAAuthor | null;
};

export type QuestionDetail = Question & { answers: Answer[] };

/** Open community questions. Public — no token needed. */
export async function getQuestions(
  productId?: string,
  options: { q?: string; limit?: number } = {},
): Promise<Question[] | null> {
  try {
    const params = new URLSearchParams({ limit: String(options.limit ?? 30) });
    if (productId) params.set("product_id", productId);
    // Free text, for the Questions tab on /search. The API matches the question
    // wording and the product name; a blank string is not a query, so it is
    // omitted rather than sent as an empty filter.
    const needle = options.q?.trim();
    if (needle) params.set("q", needle);
    return await apiFetch<Question[]>(`/api/v1/questions?${params}`, {
      revalidate: 60,
    });
  } catch {
    // See lib/requests.ts: null distinguishes "unreachable" from "none".
    return null;
  }
}

export async function getQuestionDetail(id: string): Promise<QuestionDetail | null> {
  try {
    return await apiFetch<QuestionDetail>(`/api/v1/questions/${id}`, {
      revalidate: 60,
    });
  } catch {
    return null;
  }
}

File: backend/app/schemas/qa.py
"""Community Q&A schemas (FR-5).

Questions are product-scoped and directed at either other buyers or the seller.
Answer-level earning (`answers.earn_eligible`) stays unwired (ADR-006 / A5).
"""

from __future__ import annotations

import uuid
from datetime import datetime
from decimal import Decimal

from pydantic import BaseModel, ConfigDict, Field, model_validator

from app.models.enums import QuestionDirectedTo


class QuestionCreate(BaseModel):
    """A question about exactly one subject: a product, or a store (0045)."""

    product_id: uuid.UUID | None = None
    seller_id: uuid.UUID | None = None
    body: str = Field(min_length=1, max_length=2000)
    #: Defaults to the seller for a store question and to other buyers otherwise.
    directed_to: QuestionDirectedTo | None = None

    @model_validator(mode="after")
    def _exactly_one_subject(self) -> QuestionCreate:
        if (self.product_id is None) == (self.seller_id is None):
            raise ValueError("A question is about a product or a store: give exactly one.")
        if self.directed_to is None:
            self.directed_to = (QuestionDirectedTo.seller if self.seller_id is not None
                                else QuestionDirectedTo.buyers)
        return self


class AnswerCreate(BaseModel):
    body: str = Field(min_length=1, max_length=4000)


class QAAuthor(BaseModel):
    """The public author fields a Q&A card needs."""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    username: str | None = None
    display_name: str | None = None
    trust_stage: int = 0
    trust_level_name: str | None = None
    # See FeedAuthor.reputation_score — the badge reads the same everywhere.
    reputation_score: Decimal = Decimal("0")


class AnswerOut(BaseModel):
    id: uuid.UUID
    answer_id: str | None = None
    body: str
    is_best_answer: bool = False
    is_first_responder: bool = False
    helpful_votes: int = 0
    created_at: datetime
    responder: QAAuthor | None = None
    #: True when the store's claimed owner wrote this, on a store question.
    is_seller_answer: bool = False
    #: The store's name, set only on a seller answer, so it is shown as the store.
    seller_name: str | None = None


class QuestionOut(BaseModel):
    id: uuid.UUID
    question_id: str | None = None
    #: Null for a question asked of a store.
    product_id: uuid.UUID | None = None
    product_name: str | None = None
    seller_id: uuid.UUID | None = None
    seller_name: str | None = None
    body: str
    directed_to: QuestionDirectedTo
    best_answer_id: uuid.UUID | None = None
    answer_count: int = 0
    created_at: datetime
    asker: QAAuthor | None = None


class QuestionDetailOut(QuestionOut):
    answers: list[AnswerOut] = []
